libs/wingc/src/jsify.rs

```diff
@@ -274,26 +274,59 @@ impl<'a> JSifier<'a> {
 					// This should only happen in the case of `any`, which are almost certainly JSII imports.
 					true
 				};
+				let is_abstract = if let Some(cls) = expression_type.unwrap().as_class_or_resource() {
+					cls.is_abstract
+				} else {
+					false
+				};
 
-				// If this is a resource then add the scope and id to the arg list
-				if is_resource {
-					format!(
-						"new {}({})",
-						self.jsify_type(class),
-						self.jsify_arg_list(
-							&arg_list,
-							Some("this"),
-							Some(&format!("{}", obj_id.as_ref().unwrap_or(&self.jsify_type(class)))),
-							should_case_convert,
-							phase
-						)
-					)
+				// if we have an FQN, we emit a call to the "new" (or "newAbstract") factory method to allow
+				// targets and plugins to inject alternative implementations for types. otherwise (e.g.
+				// user-defined types), we simply instantiate the type directly (maybe in the future we will
+				// allow customizations of user-defined types as well, but for now we don't).
+
+				let ctor = self.jsify_type(class);
+
+				let scope = if is_resource {
+					Some("this")
 				} else {
-					format!(
-						"new {}({})",
-						self.jsify_type(&class),
-						self.jsify_arg_list(&arg_list, None, None, should_case_convert, phase)
-					)
+					None
+				};
+
+				let id = if is_resource {
+					Some(obj_id.as_ref().unwrap_or(&ctor).as_str())
+				} else {
+					None
+				};
+
+				let args = self.jsify_arg_list(
+					&arg_list,
+					scope,
+					id,
+					should_case_convert,
+					phase
+				);
+
+				let fqn = if is_resource {
+					expression_type
+						.expect("expected expression")
+						.as_resource()
+						.expect("expected resource")
+						.fqn
+						.clone()
+				} else {
+					None
+				};
+
+				if ! is_resource || fqn.is_none() {
```

Comment:
Any rustaceans in the room to help me write this more idiomatically?

Reply:
@eladb (you nerd sniped me)

One idea I had was to try flipping the conditional, and using `if let` with an "&&" condition -- but it turns out this syntax isn't stable yet so it's possible so you will get a compiler warning 😅:

```rust
if let Some(fqn) = fqn && is_resource {
  // this.node.root ...
} else {
  // new ...
}
```

So the other way I think you can do this to avoid the extra `expect` is to combine the values into a tuple and then match:

```rust
if let (Some(fqn), true) = (fqn, is_resource) {
	if is_abstract {
		format!("this.node.root.newAbstract(\"{}\",{})", fqn, args)
	} else {
		format!("this.node.root.new(\"{}\",{},{})", fqn, ctor, args)
	}
} else {
	format!("new {}({})", ctor, args)
}
```

Reply:
Is there currently a reason to check `is_resource`? I understand that in the future that check would be needed, but based on the code above if is_resource is true then fqn is always `Some` at this point.

so just:

```rust
if let Some(fqn) = fqn {
	if is_abstract {
		format!("this.node.root.newAbstract(\"{}\",{})", fqn, args)
	} else {
		format!("this.node.root.new(\"{}\",{},{})", fqn, ctor, args)
	}
} else {
	format!("new {}({})", ctor, args)
}
```

If you want to keep the check in for the future, then tbh I think what you put initially is perfectly legible. You could even instead chuck an assert for `is_resource == true` inside the `if let` if you wanted to make sure this gets updated later.

Reply:
I'll merge it for now and will revisit later. (just don't have time to deal with it right now and it's a big enough change that's worth not lingering on...)